src/span-data.ts

```diff
@@ -42,36 +41,49 @@ function randomSpanId() {
   return parseInt(spanRandomBuffer().toString('hex'), 16).toString();
 }
 
-export class SpanData implements SpanDataInterface {
+/**
+ * Represents a real trace span.
+ */
+export abstract class BaseSpanData implements SpanData {
   readonly span: TraceSpan;
+  readonly trace: Trace;
+  abstract readonly type: SpanDataType;
 
   /**
    * Creates a trace context object.
    * @param trace The object holding the spans comprising this trace.
    * @param spanName The name of the span.
-   * @param parentSpanId The id of the parent span, 0 for root spans.
-   * @param isRoot Whether this is a root span.
-   * @param skipFrames the number of frames to remove from the top of the stack.
-   * @constructor
+   * @param parentSpanId The ID of the parent span, or '0' to specify that there
+   *                     is none.
+   * @param skipFrames the number of frames to remove from the top of the stack
+   *                   when collecting the stack trace.
    */
   constructor(
-      readonly trace: Trace, spanName: string, parentSpanId: string,
-      private readonly isRoot: boolean, skipFrames: number) {
-    spanName =
-        traceUtil.truncate(spanName, Constants.TRACE_SERVICE_SPAN_NAME_LIMIT);
-    this.span = new TraceSpan(spanName, randomSpanId(), parentSpanId);
-    trace.spans.push(this.span);
+      trace: Trace, spanName: string, parentSpanId: string,
```

Comment:
Why did you change the inline declaration of the `trace` property?

Reply:
It looks cleaner. I forgot that the TS style guide prefers inline... will change.

Reply:
Avoiding style changes in big PRs also makes the reviewers' life easy. If you really want them, put them in a standalone commit for ease of reviewing.